From f1a9d76c619064e453674849c5c37e57c4368dc6 Mon Sep 17 00:00:00 2001
From: "Aaron T. Myers" <atm@apache.org>
Date: Mon, 20 Jun 2016 13:46:11 -0700
Subject: [PATCH 1952/2848] HDFS-10423. Increase default value of httpfs
 maxHttpHeaderSize. Contributed by Nicolae Popa.

(cherry picked from commit aa1b583cf99d1a7cfe554d1769fc4af252374663)

Conflicts:
	hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/libexec/httpfs-config.sh

(cherry picked from commit 0f616cefeb3a946a9a359cf2234b154cfed47072)

Conflicts:
	hadoop-hdfs-project/hadoop-hdfs-httpfs/src/site/markdown/ServerSetup.md.vm

Change-Id: Ib1d22162f5a12aa297584cd89e695d3cace7dfe8
---
 .../hadoop-hdfs-httpfs/src/main/conf/httpfs-env.sh |    4 ++++
 .../src/main/libexec/httpfs-config.sh              |    7 +++++++
 .../hadoop-hdfs-httpfs/src/main/tomcat/server.xml  |    1 +
 .../src/main/tomcat/ssl-server.xml                 |    1 +
 .../src/site/apt/ServerSetup.apt.vm                |    4 ++++
 5 files changed, 17 insertions(+)

diff --git a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/conf/httpfs-env.sh b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/conf/httpfs-env.sh
index a2701d4..f19f706 100644
--- a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/conf/httpfs-env.sh
+++ b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/conf/httpfs-env.sh
@@ -44,6 +44,10 @@
 #
 # export HTTPFS_SSL_ENABLED=false
 
+# The maximum size of Tomcat HTTP header
+#
+# export HTTPFS_MAX_HTTP_HEADER_SIZE=65536
+
 # The location of the SSL keystore if using SSL
 #
 # export HTTPFS_SSL_KEYSTORE_FILE=${HOME}/.keystore
diff --git a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/libexec/httpfs-config.sh b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/libexec/httpfs-config.sh
index 94a97b8..0810b3d 100644
--- a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/libexec/httpfs-config.sh
+++ b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/libexec/httpfs-config.sh
@@ -164,6 +164,13 @@ else
   print "Using   HTTPFS_SSL_KEYSTORE_PASS:     *REDACTED*"
 fi
 
+if [ "${HTTPFS_MAX_HTTP_HEADER_SIZE}" = "" ]; then
+  export HTTPFS_MAX_HTTP_HEADER_SIZE=65536
+  print "Setting HTTPFS_MAX_HTTP_HEADER_SIZE:     ${HTTPFS_MAX_HTTP_HEADER_SIZE}"
+else
+  print "Using   HTTPFS_MAX_HTTP_HEADER_SIZE:     ${HTTPFS_MAX_HTTP_HEADER_SIZE}"
+fi
+
 if [ "${CATALINA_BASE}" = "" ]; then
   export CATALINA_BASE=${HTTPFS_HOME}/share/hadoop/httpfs/tomcat
   print "Setting CATALINA_BASE:       ${CATALINA_BASE}"
diff --git a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/tomcat/server.xml b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/tomcat/server.xml
index a425bdd..67f2159 100644
--- a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/tomcat/server.xml
+++ b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/tomcat/server.xml
@@ -71,6 +71,7 @@
     -->
     <Connector port="${httpfs.http.port}" protocol="HTTP/1.1"
                connectionTimeout="20000"
+               maxHttpHeaderSize="${httpfs.max.http.header.size}"
                redirectPort="8443"/>
     <!-- A "Connector" using the shared thread pool-->
     <!--
diff --git a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/tomcat/ssl-server.xml b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/tomcat/ssl-server.xml
index 287c9f5..92bee15 100644
--- a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/tomcat/ssl-server.xml
+++ b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/tomcat/ssl-server.xml
@@ -70,6 +70,7 @@
          described in the APR documentation -->
     <Connector port="${httpfs.http.port}" protocol="HTTP/1.1" SSLEnabled="true"
                maxThreads="150" scheme="https" secure="true"
+               maxHttpHeaderSize="${httpfs.max.http.header.size}"
                clientAuth="false" sslEnabledProtocols="TLSv1,TLSv1.1,TLSv1.2,SSLv2Hello"
                keystoreFile="${httpfs.ssl.keystore.file}"
                keystorePass="${httpfs.ssl.keystore.pass}"/>
diff --git a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/site/apt/ServerSetup.apt.vm b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/site/apt/ServerSetup.apt.vm
index 878ab1f..50a5b48 100644
--- a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/site/apt/ServerSetup.apt.vm
+++ b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/site/apt/ServerSetup.apt.vm
@@ -102,6 +102,8 @@ Transfer-Encoding: chunked
 
   Tomcat logs are also preconfigured to go to HttpFS's <<<logs/>>> directory.
 
+  HttpFS default value for the maxHttpHeaderSize parameter in Tomcat's `server.xml` is set to 65536 by default.
+
   The following environment variables (which can be set in HttpFS's
   <<<conf/httpfs-env.sh>>> script) can be used to alter those values:
 
@@ -111,6 +113,8 @@ Transfer-Encoding: chunked
 
   * HTTPFS_LOG
 
+  * HTTPFS\_MAX\_HTTP\_HEADER\_SIZE
+
 * HttpFS Configuration
 
   HttpFS supports the following {{{./httpfs-default.html}configuration properties}}
-- 
1.7.9.5

